src/contexts/Toast/ToastProvider.tsx

```diff
@@ -0,0 +1,115 @@
+import React, { createContext, useContext, useEffect, useState, useRef } from 'react'
+import { Message, Transition } from 'semantic-ui-react'
+import { ToastProps, MessageProps, Position, bottomLeft } from './types'
+import { getStyleProps } from './helpers'
+import { nanoid } from 'nanoid'
+
+const TRANSITION_DURATION = 350 // ms
+
+const setTimeout = window.setTimeout // To ensure return type is number
+interface ToastProviderValue {
+  showToast: (...state: Partial<ToastProps>[]) => void
+  updateDefaults: (state: Partial<ToastProps>) => void
+}
+
+const ToastProviderContext = createContext<ToastProviderValue>({
+  showToast: () => {},
+  updateDefaults: () => {},
+})
+
+export function ToastProvider({ children }: { children: React.ReactNode }) {
+  const [toastDefaults, setToastDefaults] = useState<ToastProps>({
+    title: '',
+    text: '',
+    style: 'basic',
+    timeout: 5000,
+    clickable: true,
+    showCloseIcon: true,
+    position: bottomLeft,
+    uid: '',
+  })
+  const [toasts, setToasts] = useState<ToastProps[]>([])
+
+  const updateDefaults = (newDefaults: Partial<ToastProps>) => {
+    setToastDefaults((prevDefaults) => ({ ...prevDefaults, ...newDefaults }))
+  }
+
+  const showToast = (newToast: Partial<ToastProps> = {}) => {
+    setToasts((prevState) => [...prevState, { ...toastDefaults, ...newToast, uid: nanoid(8) }])
+  }
+
+  const removeToast = (uid: string) => {
+    setToasts((prevState) => prevState.filter((toast) => toast.uid !== uid))
+  }
+
+  return (
+    <ToastProviderContext.Provider
+      value={{
+        showToast,
+        updateDefaults,
+      }}
+    >
+      <div id="toast-container">
+        {Object.entries(Position).map(([key, positionClass]) => (
+          <div className={`block ${positionClass}`} key={key}>
+            {toasts
+              .filter((toast) => toast.position === Position[key as keyof typeof Position])
+              .map((toast) => (
+                <Toast toast={toast} removeToast={() => removeToast(toast.uid)} key={toast.uid} />
+              ))}
+          </div>
+        ))}
+      </div>
+      {children}
+    </ToastProviderContext.Provider>
+  )
+}
+
+export const useToast = (toastSettings: Partial<ToastProps> = {}) => {
+  const { showToast, updateDefaults } = useContext(ToastProviderContext)
+  useEffect(() => {
+    updateDefaults(toastSettings)
+  }, [])
+
+  return showToast
+}
+
+export const Toast = ({ toast, removeToast }: { toast: ToastProps; removeToast: () => void }) => {
+  const [visible, setVisible] = useState(false)
+  const timerId = useRef<number>(0)
+  const { timeout } = toast
+
+  useEffect(() => {
+    setTimeout(() => setVisible(true), 50)
```

Comment:
I don't undertand what is this `setTimeout` this part seems very confusing and hard to follow what the code here is doing. I can undertand that's how the Toast will be removed after the transition_duration ends, but why do you need to call setTimeout here that many times?

Reply:
The first timeout is for the `<Transition>` component - the timeout makes the Toast start invisible, then become visible, which is how the "Transition" component can apply the nice "fade-in" effect. If we start with it visible, you don't get the effect.

(And the 2nd timeout is obviously to close the Toast)

I can see how that's not obvious though, so have added a comment above the first one: 7ab8c456